File: 1228-MinimumCostTreeFromLeafValues/1228-MinimumCostTreeFromLeafValues.py
# Last updated: 08/07/2026, 20:56:22
class Solution:
    def mctFromLeafValues(self, arr: List[int]) -> int:
        stack = [float('inf')]
        result = 0

        for num in arr:
            while stack[-1] <= num:
                mid = stack.pop()
                result += mid * min(stack[-1], num)

            stack.append(num)

        while len(stack) > 2:
            result += stack.pop() * stack[-1]

        return result

File: 1228-MinimumCostTreeFromLeafValues/1228-MinimumCostTreeFromLeafValues.java
// Last updated: 09/07/2026, 10:36:33
class Solution {
    public int mctFromLeafValues(int[] arr) {
        Stack<Integer>st=new Stack<>();
        st.add((int)1e9);
        int ans=0;
        for(int a:arr){
            while(st.peek()<=a){
                ans+=st.pop()*Math.min(st.peek(),a);
            }
            st.push(a);
        }
        while(st.size()>2){
            ans+=st.pop()*st.peek();
        }
        return ans;
    }
}